ops/relation.py

```diff
@@ -0,0 +1,548 @@
+                consumed, provider_data))
+            self.on.invalid.emit(provides)
+
+    def on_provider_broken(self, event):
+        """Inform consumer charm that provider relation no longer exists
+
+        This method raises a :class:`ProviderBroken` event in respose to
+        a relation broken event.
+
+        Args:
+            event: an event object
+        """
+        logger.debug("Provider Broken : {}".format(event))
+        if not self.multi_mode:
+            self.stored.relation_id = None
+        self.on.broken.emit()
+
+    def validate_provider(self, event):
+        """Check provider and consumer compatibility
+
+        This method validates provider consumer compatibility using
+        data that is already available in the application relation
+        bucket.
+
+        Args:
+            event: an event object
+        """
+        logger.debug("Validating provider(s) : {}".format(event))
+        consumed = self.consumes
+
+        for relation in self.framework.model.relations[self.name]:
+            data = self.provider_data(relation.id)
+            if data:
+                try:
+                    provides = data['provides']
+                except KeyError:
+                    continue
+
+            requirements_met = self.meets_requirements(provides, consumed)
+            if requirements_met:
+                self.on.available.emit(data)
+            else:
+                logger.error('Provider non longer compatible, Need {}, have {}'.format(
+                    consumed, data))
+                self.on.invalid.emit(data)
+
+    def meets_requirements(self, provides, consumes):
+        """Check if provider and consumer are compatible.
+
+        Args:
+            provides: a dictionary with a single key value pair. The key
+                is a string naming the resource provided. The value is a
+                string given the version of the provided resource.
+            consumes: a dictionary with zero or more key value paris. Each
+                key is a string name of a resource that is acceptable. The
+                corresponding value is a string representing an acceptable
+                version specification. The version specification can be in any
+                format that is compatible with the
+                [semver](https://pypi.org/project/semver/) Python package.
+
+        Returns:
+            bool: True if the producer and consumer specification are
+            compatible.
+        """
+        assert(len(provides) == 1)
+        provided = tuple(provides.items())[0]
+        for required in consumes.items():
+            if self.is_compatible(provided, required):
+                return True
+        return False
+
+    def is_compatible(self, has, needs):
+        """Is a producer and consumer specification compatible.
+
+        Args:
+            has: a dictionary with a single key value pair. The key
```

Comment:
`provided = tuple(provides.items())[0]` is not a dictionary with a single key value pair:

```
>>> x = {'a': 1}
>>> tuple(x.items())[0]
('a', 1)
```

Given normalized_type assumes [0] indexing, I'm guessing it does assume that it is a tuple.

Reply:
`meets_requirements` check compatibility between the `provides` and `consumes` specification by iterating over all possibilities. So for example `provides` can be `{"mysql": ">=6.0"}` and `consumes` can be `{"mysql": "6.4", "mariadb": "7.0"}. `is_compatible` check compatibility between the single provides pair (tuple of type and version) and one of the consumers pair (also tuple type and version). 

I hope this answers your question.